libbpf-tools/fsdist.c

```diff
@@ -0,0 +1,373 @@
+		bpf_program__set_autoload(obj->progs.nfs_getattr_fentry, true);
+		bpf_program__set_autoload(obj->progs.nfs_getattr_fexit, true);
+	} else {
+		bpf_program__set_autoload(obj->progs.nfs_file_read_entry, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_read_return, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_write_entry, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_write_return, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_open_entry, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_open_return, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_sync_entry, true);
+		bpf_program__set_autoload(obj->progs.nfs_file_sync_return, true);
+		bpf_program__set_autoload(obj->progs.nfs_getattr_entry, true);
+		bpf_program__set_autoload(obj->progs.nfs_getattr_return, true);
+	}
+}
+
+static void enable_xfs_probes(struct fsdist_bpf *obj)
+{
+	if (fentry_exists("xfs_file_read_iter", "xfs")) {
+		bpf_program__set_autoload(obj->progs.xfs_file_read_fentry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_read_fexit, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_write_fentry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_write_fexit, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_open_fentry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_open_fexit, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_sync_fentry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_sync_fexit, true);
+	} else {
+		bpf_program__set_autoload(obj->progs.xfs_file_read_entry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_read_return, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_write_entry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_write_return, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_open_entry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_open_return, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_sync_entry, true);
+		bpf_program__set_autoload(obj->progs.xfs_file_sync_return, true);
+	}
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct fsdist_bpf *skel;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	skel = fsdist_bpf__open();
+	if (!skel) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	skel->rodata->target_pid = target_pid;
+	skel->rodata->in_ms = timestamp_in_ms;
+
+	disable_autoload(skel);
+	switch (fs_kind) {
+	case BTRFS:
+		enable_btrfs_probes(skel);
+		break;
+	case EXT4FS:
+		enable_ext4fs_probes(skel);
+		break;
+	case NFS:
+		enable_nfs_probes(skel);
+		break;
+	case XFS:
+		enable_xfs_probes(skel);
+		break;
+	}
+
```

Comment:
So I was thinking to avoid so many BPF programs and hard-coding all those fs-specific function names.

Think about this approach.

1. You define "generic" fentry/fexit programs for each possible type of fs_op. You don't even have to specify real function names in .bpf.c, because we are going to override them at runtime. Similarly, you define BPF_KRETPROBE for each such operation, again with fake attach function names. You can probably re-use the same BPF_KPROBE, btw, given they do the same thing that's not operation-specific.

2. On user-space side, we have a table of supported file systems and operations (and some operations might not be supported at all, like NFS has more stuff than other FSs):
```
struct {
    const char *fs;
    const char *op_funcs[MAX_OP];
} fs_configs[] = {
  { "btrfs", {
      [OPEN] = "ext4_file_open",
      [FSYNC] = "ext4_sync_file",
      [GETATTR] = NULL, /* not supported */
  },
  { "nfs", { ... } },
  /* and so on */
}
```

3. At runtime you search for your desired FS, iterate over supported operations and see which kernel function has to be traced. You do fentry_exists() check and either pick, say, generic fentry_open/fexit_open pair (for OPEN operation) or disable them and pick kprobe/kretrprobe. For fentry/fexit you have to use bpf_program__set_attach_target() call to specify exact function you want to eventually attach to. That has to happen before skeleton load. For kprobe/kretprobe, you'll do bpf_program__attach_kprobe() after skeleton loads (so you'll have two loops to handle both fentry and kprobe, that's a mild complication).

And that's it. Your BPF side will have just minimum necessary set of BPF programs. Your user-space will be easily extensible to new FSs by just adding "config". Done.

4. Let's also add the logic of checking with which name fsdist was exec()'ed. If it is, say, ext4dist, then just automatically pretend if `-t ext4` was provided. Then we can just have symlinks from ext4dist to fsdist to preserve the behavior of dedicated ext4dist, btrfsdist, xfsdist, and so on.

Reply:
Ah, thanks. Nice idea, will do in this way. :)

(When I implemented this, I found myself didn't follow the DRY rule, but couldn't figure out a good way to organize the code.)

Reply:
> checking with which name fsdist was exec()'ed

Do you mean `basename(argv[0])` ? I don't know how to do that. :facepalm:

Reply:
yeah, something like that. Doesn't it work? I've seen perf tool doing something like that.